protocol/consistencychecks.go

```diff
@@ -0,0 +1,345 @@
+// state of a CONIKS client. This includes the latest epoch
+// and SignedTreeRoot value, as well as directory's policies
+// (e.g., whether the TemporaryBinding extension is being used or not)
+// The client should create a new ConsistencyChecks instance only once,
+// when she registers her binding with a ConiksDirectory.
+// This ConsistencyChecks instance then will be used to verify
+// the returned responses from the ConiksDirectory.
+type ConsistencyChecks struct {
+	// Timestamp indicates the time when the SavedSTR updated.
+	Timestamp m.Timestamp
+	SavedSTR  *m.SignedTreeRoot
+
+	// extensions settings
+	useTBs bool
+	TBs    map[string]*TemporaryBinding
+
+	// signing key
+	signKey sign.PublicKey
+}
+
+// NewCC creates an instance of ConsistencyChecks using
+// the pinning directory's STR at epoch 0 or
+// the consistency state read from a persistent storage.
+func NewCC(savedSTR *m.SignedTreeRoot, useTBs bool, signKey sign.PublicKey) *ConsistencyChecks {
+
+	// TODO: see #110
+	if !useTBs {
+		panic("[coniks] Currently the server is forced to use TBs")
+	}
+
+	cc := &ConsistencyChecks{
+		Timestamp: m.Timestamp(time.Now().Unix()),
+		SavedSTR:  savedSTR,
+		useTBs:    useTBs,
+		signKey:   signKey,
+	}
+
+	if useTBs {
+		cc.TBs = make(map[string]*TemporaryBinding)
+	}
+	return cc
+}
+
+// UpdateConsistency verifies the directory's response for a request.
+// It will also update the state if the requestType is
+// either RegistrationType or MonitoringType.
+// It first verifies the directory's returned status code of the request.
+// If the status code is not in the ErrorResponses array, it means
+// the directory has successfully handled the request.
+// The verifier will then verify the consistency state of the response.
+// This will panic if it is called with an int
+// which isn't a valid/known request-type.
+func (cc *ConsistencyChecks) UpdateConsistency(requestType int, msg *Response,
+	uname string, key []byte) ErrorCode {
+	if ErrorResponses[msg.Error] {
+		return msg.Error
+	}
+
+	var str *m.SignedTreeRoot
+	var err error
+
+	switch requestType {
+	case RegistrationType:
+		str, err = cc.verifyRegistration(requestType, msg, uname, key)
+
+	case KeyLookupType:
+		err = cc.verifyKeyLookup(requestType, msg, uname, key)
+
+	case MonitoringType:
+	case KeyLookupInEpochType:
+	default:
+		panic("[coniks] Unknown request type")
+	}
+
+	if err == Passed &&
+		requestType == RegistrationType || requestType == MonitoringType {
+		// update the state
+		cc.Timestamp = m.Timestamp(time.Now().Unix())
+		cc.SavedSTR = str
+	}
+
+	return err.(ErrorCode)
+}
+
+func (cc *ConsistencyChecks) verifyRegistration(requestType int,
+	msg *Response, uname string, key []byte) (*m.SignedTreeRoot, error) {
+
+	df, ok := msg.DirectoryResponse.(*DirectoryProof)
+	if !ok {
+		return nil, ErrorMalformedDirectoryMessage
+	}
+
+	ap := df.AP
+	str := df.STR
+
+	interval := m.Timestamp(time.Now().Unix()) - cc.Timestamp
```

Comment:
Hmm, this codeblock feels misplaced.  Transitioning between epochs on the client side hasn't been something we've discussed much yet, have we?

I think we always want `verifyRegistration` and `verifyKeyLookup` to just do `verifySTR`, and the current `UpdateConsistency` should just do verifications, and be renamed to `CheckConsistency`, or whatever, as it previously was.

Then we need a new function `UpdateSTR` that the client would run before any consistency check, when it made any request.  I know this contradicts what I said previously about lookups not changing state, but I hadn't considered time skew, and that a lookup may happen after an epoch ended but before our clock tells us to do a monitoring request.  Tor depends on NTP and we probably need to as well.

`UpdateSTR` would then do then `verifySTRConsistency`, if necessary.  I'm starting to see that you've already thought about all this, just the organization of the code obscured it from me.  Sorry to be slow.